doc/tutorials/debug.rst

```diff
@@ -0,0 +1,371 @@
+
+First, we can get the USB controller BDF number (0:15.0) through the
+following command::
+
+   lspci | grep "USB controller"
+
+
+.. figure:: images/debug_image19.png
+   :align: center
+
+   USB controller BDF information
+
+Second, we use the ``pt`` command and use this BDF number
+to find the interrupt vector (VEC) "0x31".
+
+.. figure:: images/debug_image20.png
+   :align: center
+
+   USB controller interrupt information
+
+Finally we use the ``int`` command and use this interrupt vector
+to find the interrupt number (909) on CPU3.
+
+.. figure:: images/debug_image21.png
+   :align: center
+
+   USB controller interrupt number information
+
+ACRN Log
+********
+
+ACRN log provides console log and acrnlog for a user to analyze.
+We can use console log to debug directly, while acrnlog is a userland tool
+used to capture a ACRN hypervisor log.
+
+console log/acrn log
+====================
+
+We can set console log level and acrnlog level through::
+
+   loglevel [console_loglevel] [mem_loglevel]
+
+For example, we add the following code into function ``shell_cmd_help`` of
+the source file ``acrn-hypervisor/hypervisor/debug/shell.c``..
+
+.. figure:: images/debug_image22.png
+   :align: center
+
+   shell_cmd_help added information
+
+We set console log level to 5, and acrnlog level to 2 through the
+command::
+
+  loglevel 5 2
+
+Then we input "help" into ACRN console, and find the log as follows.
+
+.. figure:: images/debug_image23.png
+   :align: center
+
+   console log information
+
+Then we use the command::
+
+   sos_console
+
+to switch to the SOS console the use the commands::
+
+   systemctl enable acrnlog
+   reboot
+
+to enable the acrnlog and reboot the SOS. Next we set the console and
+acrnlog levels with::
+
+   loglevel 5 2
+
+and run the command::
+
+   help
+
+Finally we get into the SOS and use the command::
+
+   cat /tmp/acrnlog/acrnlog_cur.0
+
+on the console, and we'll see the following log:
+
+.. figure:: images/debug_image24.png
+   :align: center
+
+   acrnlog information
+
+ACRN Trace
+**********
+
+ACRN trace is a tool running on the Service OS (SOS) to capture trace
+data.
+
+trace setting
+=============
+
+First we add the following event id into
+``acrn-ebtool/acrn-hypervisor/hypervisor/include/debug/trace.h``:
+
+.. figure:: images/debug_image25.png
+   :align: center
+
+   trace event id
+
+Then we add the following format to
+``acrn-hypervisor/tools/acrntrace/scripts/formats``:
+
+.. figure:: images/debug_image1.png
+   :align: center
+
+   acrntrace formatted information
+
+We add trace into function ``emulate_io`` in
+``acrn-hypervisor/hypervisor/arch/x86/guest/io_emul.c``.
+
+.. figure:: images/debug_image2.png
+   :align: center
+
+   inserted trace information
+
+trace collection
+================
+
+In the SOS console we run::
+
+   acrntrace -c
+
+to generate acrntrace data into the current directory.
+
+.. figure:: images/debug_image3.png
+   :align: center
+
+   trace collection
+
+trace analysis
```

Comment:
Can this be merged/combined with https://projectacrn.github.io/latest/tools/acrntrace/README.html